### ISSUE TITLE
Add way to easily test server methods

### ISSUE THREAD

**Issue by L0g4n** [hyperium/tonic#727]
### Motivation

**tl;dr; Testing is important, but tonic currently makes it hard perform proper integration testing**

I want to be able to basically perform proper unit testing / integration testing of my implemented server methods. That means, in my tests I want to connect to the grpc server, execute some methods, and then assert some results. Similar to performing REST API testing.

However, for that to work transparently I would need to spawn the grpc server which would block the execution currently (and then I could not perform the client method calls anymore). I think for that to work I would need to spawn the server on a separate thread, but I am not that familiar with rust yet and know if this would even work (can I even await futures executed on another OS thread? Or do I need a separate tokio runtime for that?)

As a workaround I execute the server in the background before running my tests. This works, though not being ideal. I think it would be good if testing would become a first class citizen in this crate, similar to the [test](https://docs.rs/warp/0.3.1/warp/test/index.html) module of the warp crate.

**Comment by davidpdrsn** [comment]
Yeah you're right that testing tonic server's is a bit more complicated than regular REST servers, mostly because of the code generation. You cannot "just" make an HTTP request and fire it off. It needs to match the schema.

The recommended way if to spawn a server on a background task and communicate with it like you would any other TCP server. Like we do [here](https://github.com/hyperium/tonic/blob/master/tests/integration_tests/tests/timeout.rs). I use this technique at work and its worked very well.

If you don't want to run a server in the background then you can also mock the transport using `tokio::io::duplex` like shown [here](https://github.com/hyperium/tonic/blob/master/examples/src/mock/mock.rs).

Finally, if you don't care about testing the transport stuff like tower middleware, you can create an instance of your server struct and call the methods on it directly. The ones from the generated async trait that you implement.

Would one of these options work for you?

**Comment by L0g4n** [comment]
>Yeah you're right that testing tonic server's is a bit more complicated than regular REST servers, mostly because of the code generation. You cannot "just" make an HTTP request and fire it off. It needs to match the schema.

Sorry if I maybe communicated a bit unclear there, I know of course its a bit more complicated than just firing off a HTTP request. Of course I meant from a client which has the proper methods generated in them.

The three options you mentioned sound all promising, they all sound reasonable at the moment. What do you think makes more sense? Having the test environment near to the actual real world case or mocking the whole transport layer?

**Comment by davidpdrsn** [comment]
There should be no difference between using `tokio::io::duplex` or using a `tokio::net::TcpStream`. Both should give you equal guarantees that things will work in production as well. Tonic actually doesn't know what its writing to, just that it implements `AsyncRead` and `AsyncWrite`.

So which method you pick depends on your setup. I think I would go for `tokio::io::duplex` if I was starting from scratch but I use `tokio::net::TcpStream` at work because it makes other parts of our setup simpler.

**Comment by davidpdrsn** [comment]
`tokio::io::duplex` is also a bit easier since you don't need to worry about creating tcp streams and ports and all that.

**Comment by L0g4n** [comment]
So, the `tokio::io::duplex` provided worked great, thanks for that  (this example reminded me to upgrade to tonic 0.5 btw, since the `Connected` traits interface changed) 👍.

Although I have one question regarding this snippet:
```rust
let channel = Endpoint::try_from("http://[::]:50051")?
        .connect_with_connector(service_fn(move |_: Uri| {
            let client = client.take().unwrap();

            async move { Ok::<_, std::io::Error>(mock::MockStream(client)) }
        }))
        .await?;
```
Where does the destination ip `"http://[::]:50051"` come from? Or is this just a placeholder and is not used anyway since we then use a custom connector with `connect_with_connector``

**Comment by davidpdrsn** [comment]
Yeah its just a placeholder. Its not used when you use `connect_with_connector`. We're aware of it but don't have a good solution to it yet. Suggestions are welcome 😊

**Comment by L0g4n** [comment]
OK, I consider this issue then closed with those examples in place.
